### PR TITLE
Merge da-ghc-8.8.1 branch into master

### PR DESCRIPTION
**Pull request by cocreature**
Sorry for the messed up history. The branches are so far apart that I’m not sure how to make this not ugly.

I’m also fairly sure that there are still some ifdefs missing. I’ll work those out with CI.

### CODE REVIEW

**Comment by shayne-fletcher-da** [comment]
I think the build failure on master is 

> I’m also fairly sure that there are still some ifdefs missing. I’ll work those out with CI.

I think the `cpp-options` in `Ghclibgen.hs` are going to need some adjustments according to flavor e.g. `-DHAVE_INTERPRETER` on master, `-DGHCI` on the 8.8.1 branches.

**Comment by shayne-fletcher-da** [comment]
[This commit](https://github.com/digital-asset/ghc-lib/commit/83f8e15ac1a70db8adb9b1a21fb3c27c6c06c37e) highlights the differences in the `cpp-options`.

File: CI.hs [hunk 1]
```diff
@@ -1,9 +1,11 @@
+#!/usr/bin/env stack --resolver lts-14.3 runhaskell --package extra --package optparse-applicative --
```
[the remark this answers is not in the document]
Comment (cocreature) [reply CI.hs]: Should work fine afaict.

File: CI.hs [hunk 2]
```diff
@@ -70,14 +161,17 @@ main = do
...
+        then unlines ["flags: {mini-compile: {daml-unit-ids: true}}"]
+        else ""
 
     -- Separate the two library build commands so they are
     -- independently timed. Note that optimizations in these builds
     -- are disabled in stack.yaml via `ghc-options: -O0`.
     cmd "stack build ghc-lib-parser --no-terminal --interleaved-output"
     cmd "stack build ghc-lib --no-terminal --interleaved-output"
-    cmd "stack build mini-hlint mini-compile strip-locs --no-terminal --interleaved-output"
+    cmd "stack build mini-hlint mini-compile --no-terminal --interleaved-output"
```
[the remark this answers is not in the document]
Comment (cocreature) [reply CI.hs]: Good catch, fixed.

File: stack.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # Use the same resolver as GHC/Hadrian to increase caching
-resolver: lts-13.21
+resolver: lts-14.3
```
[the remark this answers is not in the document]
Comment (cocreature) [reply stack.yaml]: Oh I missed the comment that this was selected to increase caching. Not sure we get that much caching anyway tbh as hadrian compiles against a different `Cabal` but I can change it back.

**Comment by cocreature** [comment]
Good catch with the CPP messed up the merge there, fixed.

File: CI.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (shayne-fletcher-da) [reply CI.hs]: ✔️

File: CI.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shayne-fletcher-da) [reply CI.hs]: ✔️

File: stack.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (shayne-fletcher-da) [reply stack.yaml]: I'm not fussed.

**Pull request merged**